sympy/solvers/ode/systems.py

```diff
@@ -1076,11 +1076,78 @@ def _preprocess_eqs(eqs):
     return processed_eqs
 
 
-# For now, this function returns a simple output, later it will
-# be capable of dividing the system of ODEs into strongly and
-# weakly connected components.
-def _component_division(eqs, funcs):
-    return [[[eqs, funcs]]]
+def _sysode_graph(eqs):
+    funcs = [eq.lhs.args[0] for eq in eqs]
+    rhss = [eq.rhs for eq in eqs]
+
+    edges = []
+    for func, rhs in zip(funcs, rhss):
+        for other in funcs:
+            if other != func and rhs.has(other):
+                edges.append((func, other))
+    return funcs, edges
+
+
+def _is_type1(scc, t):
+    eqs, funcs = scc
+
+    try:
+        (A1, A0), b = linear_ode_to_matrix(eqs, funcs, t, 1)
+    except (ODENonlinearError, ODEOrderError):
+        return False
+
+    if _matrix_is_constant(A0, t) and b.is_zero_matrix:
+        return True
+
+    return False
+
+
+def _combine_systems(sccs, t):
+    can_combine = _is_type1(sccs[0], t)
+
+    if can_combine:
+        end_idx = 1
+        for scc in sccs[1:]:
+            eqs, funcs = scc
+            old_funcs = sccs[0][1]
+            subs_dict = {f: funcs[0] for f in old_funcs}
+            check_eqs = [eq.subs(subs_dict) for eq in eqs]
+
+            if not _is_type1((check_eqs, funcs), t):
+                break
+
+            old_eqs = sccs[0][0]
+            sccs[0] = (old_eqs + eqs, old_funcs + funcs)
+            end_idx += 1
+
+        del sccs[1:end_idx]
+
```

Comment:
@oscarbenjamin In this function, I have added a way to combine the sub-systems when they form a type 1(constant coeff homogeneous) system of ODEs. 

So, the first thing to be observed is that if the first sub-system is not of Type 1, then the other sub-systems can't be of Type 1 either. Hence, if there is exists a larger Type 1 sub-system, then it starts from the first component in a list of strongly connected components. Now, to check if the second, third, ..., etc systems are Type 1, we have to replace the dependent variables from the previous components as one of the dependent variables of this component. This makes the check a whole lot easier. If the current component is Type 1, we simply add this component with the first component and keep doing it until we reach a sub-system which is not of Type 1.

Is this logic correct?

Reply:
How does it handle a case like the following?
```
dx/dt = x**2
dy/dt = z
dz/dt = y
dw/dt = x + z
```
Here the components are `{x}, {y, z}, {w}` but you might get them in that order or you might get them in the order `{y, z}, {x}, {w}`. The component `{y, z}` is linear with constant coefficients but it might come before or after `{x}` which is nonlinear.

Reply:
Well then it would certainly be a problem. I think to avoid this, what can be done is checking the graph to see if one connected component has a directed edge to another. I would try to implement it soon.

Reply:
I have tried to fix this in the recent commit. In that, I have built a graph for strongly connected components in the form of `parents` list. This is a list of lists, each index represents the strongly connected component(scc). Then in `_combine_systems` function, I combine the systems which are of type1 and along with that keep track of its parent sccs too. If any parent scc is not type1, then this scc won't be combined. After the check, we combine the scc to the scc with the lowest index in the list of sccs to keep the topological order intact.

Reply:
That sounds reasonable.